internal/uvm/create_wcow.go

```diff
@@ -204,10 +154,138 @@ func CreateWCOW(ctx context.Context, opts *OptionsWCOW) (_ *UtilityVM, err error
+//
+func CreateWCOW(ctx context.Context, opts *OptionsWCOW) (_ *UtilityVM, err error) {
+	ctx, span := trace.StartSpan(ctx, "uvm::CreateWCOW")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+
+	if opts.ID == "" {
+		g, err := guid.NewV4()
+		if err != nil {
+			return nil, err
+		}
+		opts.ID = g.String()
+	}
+
+	span.AddAttributes(trace.StringAttribute(logfields.UVMID, opts.ID))
+	log.G(ctx).WithField("options", fmt.Sprintf("%+v", opts)).Debug("uvm::CreateWCOW options")
+
+	uvm := &UtilityVM{
+		id:                      opts.ID,
+		owner:                   opts.Owner,
+		operatingSystem:         "windows",
+		scsiControllerCount:     1,
+		vsmbDirShares:           make(map[string]*VSMBShare),
+		vsmbFileShares:          make(map[string]*VSMBShare),
+		vpciDevices:             make(map[string]*VPCIDevice),
+		physicallyBacked:        !opts.AllowOvercommit,
+		devicesPhysicallyBacked: opts.FullyPhysicallyBacked,
+	}
+
+	defer func() {
+		if err != nil {
+			uvm.Close()
+		}
+	}()
+
+	if err := verifyOptions(ctx, opts); err != nil {
+		return nil, errors.Wrap(err, errBadUVMOpts.Error())
+	}
+
+	uvmFolder, err := uvmfolder.LocateUVMFolder(ctx, opts.LayerFolders)
+	if err != nil {
+		return nil, fmt.Errorf("failed to locate utility VM folder from layer folders: %s", err)
+	}
+
+	// TODO: BUGBUG Remove this. @jhowardmsft
+	//       It should be the responsiblity of the caller to do the creation and population.
+	//       - Update runhcs too (vm.go).
+	//       - Remove comment in function header
+	//       - Update tests that rely on this current behaviour.
+	// Create the RW scratch in the top-most layer folder, creating the folder if it doesn't already exist.
+	scratchFolder := opts.LayerFolders[len(opts.LayerFolders)-1]
+
+	// Create the directory if it doesn't exist
+	if _, err := os.Stat(scratchFolder); os.IsNotExist(err) {
+		if err := os.MkdirAll(scratchFolder, 0777); err != nil {
+			return nil, fmt.Errorf("failed to create utility VM scratch folder: %s", err)
+		}
+	}
+
+	doc, err := prepareConfigDoc(ctx, uvm, opts, uvmFolder)
+	if err != nil {
+		return nil, fmt.Errorf("error in preparing config doc: %s", err)
+	}
+
+	if !opts.IsClone {
+		// Create sandbox.vhdx in the scratch folder based on the template, granting the correct permissions to it
+		scratchPath := filepath.Join(scratchFolder, "sandbox.vhdx")
+		if _, err := os.Stat(scratchPath); os.IsNotExist(err) {
+			if err := wcow.CreateUVMScratch(ctx, uvmFolder, scratchFolder, uvm.id); err != nil {
+				return nil, fmt.Errorf("failed to create scratch: %s", err)
+			}
+		}
+
+		doc.VirtualMachine.Devices.Scsi = map[string]hcsschema.Scsi{
+			"0": {
+				Attachments: map[string]hcsschema.Attachment{
+					"0": {
+						Path:  scratchPath,
+						Type_: "VirtualDisk",
+					},
+				},
+			},
+		}
+
+		uvm.scsiLocations[0][0] = &SCSIMount{
+			vm:       uvm,
+			HostPath: doc.VirtualMachine.Devices.Scsi["0"].Attachments["0"].Path,
+			refCount: 1,
+		}
+	} else {
```

Comment:
If creating a clone, it seems like there are a bunch of things we might need to validate to ensure they match the template:

e.g.
- VA-backed
- CPU cores
- Memory

Do we know the full list of what needs to match? Should we add those checks here?

Reply:
I think there are only 2/3 things that are allowed to differ from the template configuration. I think it might be easier to just make sure that if we are cloning then annotations for memory, cpu cores etc are not provided. That way we won't have to store the entire config doc of the template and then compare against it during clone time. What do you think?
In fact IIRC John had also mentioned in one of his comments that we shouldn't store the entire config doc in the registry.

Reply:
I don't have a strong opinion on whether we enforce this through checks on the config doc itself, or if we just don't allow any customization of some of these items at all for clones. As long as we ensure we aren't creating scenarios where something breaks because the user passed the wrong config.

Out of curiosity, what happens in HCS if a VM is cloned, where the template VM has 2 cores, and the clone has 4?

Reply:
I think it will be best to just not allow any other parameters in the configuration of the cloned pod/container than what is absolutely necessary. That will avoid most of these errors. I will add that validation.

If my understanding is correct starting that cloned VM will fail with some HCS error.